040-refactor-client-examples.md

```diff
@@ -0,0 +1,65 @@
+# Refactor KafkaConfig files in Strimzi Client Examples
+
+This proposal is about simplifying the `Kafka*Config` files of the Java modules of the [Strimzi Client Examples repo](https://github.com/strimzi/client-examples).
+
+## Current situation
+
+The `Kafka*Config` files of the Java modules in the [Strimzi Client Examples repo](https://github.com/strimzi/client-examples) have become quite complex since they support the configuration of several specific Kafka client properties via environment variables.
+They contain a lot of the logic for configuring the client. 
+As an example, you can have a look at the logic around setting Kafka’s `security.protocol` option which is spread across several places in the code.
+
+## Motivation
+
+The main purpose of the Client Examples is to show users how to write a simple Kafka client based on the Kafka Consumer, Producer and Streams API. 
+However, over time, we added a lot of complexity to the client code because it was being used in the System Tests at some point in time. 
+So a user who wants to learn from the Client Examples has to dig through this code to understand what and how is configured. 
+This is currently not easy because of how the configuration is generated through the Java code.
+In addition to that, we found adding Kafka client configuration updates to the `Kafka*Config` files of the Java modules in the [Strimzi Client Examples repo](https://github.com/strimzi/client-examples) to be a little more verbose and messier than necessary.
+The current method requires creating a field, getter, setter, and hard coded String per new Kafka client config field. 
+This method does not scale well and makes for a longer and messier class. 
+So for any users who want to write their own Kafka client in Java, this is neither a good inspiration nor good code to copy.
+
+## Proposal
+
+We could greatly reduce the complexity and size of the class if we standardize the naming scheme of environment variables used to configure the clients of the [Strimzi Client Examples repo](https://github.com/strimzi/client-examples)
+
+We could standardize the env vars used to configure the Kafka client properties in the following manner:
+```
+KAFKA_<NAME_OF_KAFKA_PROPERTY_DELIMITED_BY_UNDERSCORES>
+```
+For example:
+```
+KAFKA_BOOTSTRAP_SERVERS for `bootstrap.servers` property
+KAFKA_GROUP_ID for `group.id` property
+KAFKA_TOPIC for `topic` property
```

Comment:
while the first two are examples of properties to be set for Kafka clients on creation, the last one is about the "topic" that the client is going to use (send or subscribe/consume) so it's not a configuration for Kafka client creation itself.
Maybe we should use more specific prefix? Something like `KAFKA_` for configuration properties for the Kafka client and `APP_` for something more related to the higher level application? Thinking about the others like the delay, messages count and so on for example.

Reply:
You are right ... that would be better as `STRIMZI_TOPIC` for example as it is not a KAfka configuration property.

Reply:
Understood! Just updated the example with `STRIMZI_TOPIC`